### ISSUE TITLE
Subqueries issues

### ISSUE THREAD

**Issue by rkuchan** [influxdata/influxdb#7875]
## Bug report

Ubuntu 16.04, v1.2

#### Steps to reproduce:

1. Write some data
```
> insert giraffes,name=chrono leaves_consumed=12 1485298440000000000
> insert giraffes,name=graf leaves_consumed=8 1485298440000000000
> insert giraffes,name=chrono leaves_consumed=13 1485298445000000000
> insert giraffes,name=graf leaves_consumed=2 1485298445000000000
> insert giraffes,name=chrono leaves_consumed=15 1485298450000000000
> insert giraffes,name=graf leaves_consumed=4 1485298450000000000
> insert giraffes,name=chrono leaves_consumed=13 1485298455000000000
> insert giraffes,name=graf leaves_consumed=5 1485298455000000000

> select * from giraffes
name: giraffes
time                 leaves_consumed name
----                 --------------- ----
2017-01-24T22:54:00Z 12              chrono
2017-01-24T22:54:00Z 8               graf
2017-01-24T22:54:05Z 13              chrono
2017-01-24T22:54:05Z 2               graf
2017-01-24T22:54:10Z 15              chrono
2017-01-24T22:54:10Z 4               graf
2017-01-24T22:54:15Z 13              chrono
2017-01-24T22:54:15Z 5               graf
```

2. Run a query where the subquery's `GROUP BY` clause includes a time interval and tag
```
> SELECT * FROM (SELECT MEAN("leaves_consumed") FROM "giraffes" WHERE time >= '2017-01-24T22:54:00Z' AND time <= '2017-01-24T22:54:15Z' GROUP BY time(10m),"name")

ERR:
```

#### Expected behavior:

I'd expect to get either this:
```
name: giraffes
tags: name=chrono
time                 mean
----                 ----
2017-01-24T22:50:00Z 13.25

name: giraffes
tags: name=graf
time                 mean
----                 ----
2017-01-24T22:50:00Z 4.75
```

Or this:
```
name: giraffes
time                 mean   name
----                 ----   ----
2017-01-24T22:50:00Z 13.25  chrono
2017-01-24T22:50:00Z 4.75   graf
```

#### Actual behavior:

I get `ERR:`.

#### Steps to reproduce:

1. Write some data
```
> insert giraffes,name=chrono leaves_consumed=12 1485298440000000000
> insert giraffes,name=graf leaves_consumed=8 1485298440000000000
> insert giraffes,name=chrono leaves_consumed=13 1485298445000000000
> insert giraffes,name=graf leaves_consumed=2 1485298445000000000
> insert giraffes,name=chrono leaves_consumed=15 1485298450000000000
> insert giraffes,name=graf leaves_consumed=4 1485298450000000000
> insert giraffes,name=chrono leaves_consumed=13 1485298455000000000
> insert giraffes,name=graf leaves_consumed=5 1485298455000000000

> select * from giraffes
name: giraffes
time                 leaves_consumed name
----                 --------------- ----
2017-01-24T22:54:00Z 12              chrono
2017-01-24T22:54:00Z 8               graf
2017-01-24T22:54:05Z 13              chrono
2017-01-24T22:54:05Z 2               graf
2017-01-24T22:54:10Z 15              chrono
2017-01-24T22:54:10Z 4               graf
2017-01-24T22:54:15Z 13              chrono
2017-01-24T22:54:15Z 5               graf
```

2. Run a query where the subquery includes `GROUP BY *` and the main query includes `LIMIT 1`
```
> SELECT "leaves_consumed" FROM (SELECT "leaves_consumed" FROM "giraffes" GROUP BY *) LIMIT 1

name: giraffes
time                 leaves_consumed
----                 ---------------
2017-01-24T22:54:00Z 12
2017-01-24T22:54:00Z 8
```

#### Expected behavior:
Return just one point:
```
name: giraffes
time                 leaves_consumed
----                 ---------------
2017-01-24T22:54:00Z 12
```

#### Actual behavior:
I got two points. (This also happens if I `GROUP BY` a specific tag.)

#### Steps to reproduce:

1. Write some data
```
> insert giraffes,name=chrono leaves_consumed=12 1485298440000000000
> insert giraffes,name=graf leaves_consumed=8 1485298440000000000
> insert giraffes,name=chrono leaves_consumed=13 1485298445000000000
> insert giraffes,name=graf leaves_consumed=2 1485298445000000000
> insert giraffes,name=chrono leaves_consumed=15 1485298450000000000
> insert giraffes,name=graf leaves_consumed=4 1485298450000000000
> insert giraffes,name=chrono leaves_consumed=13 1485298455000000000
> insert giraffes,name=graf leaves_consumed=5 1485298455000000000

> select * from giraffes
name: giraffes
time                 leaves_consumed name
----                 --------------- ----
2017-01-24T22:54:00Z 12              chrono
2017-01-24T22:54:00Z 8               graf
2017-01-24T22:54:05Z 13              chrono
2017-01-24T22:54:05Z 2               graf
2017-01-24T22:54:10Z 15              chrono
2017-01-24T22:54:10Z 4               graf
2017-01-24T22:54:15Z 13              chrono
2017-01-24T22:54:15Z 5               graf
```

2. Run a super simple query with a subquery
```
> SELECT * FROM (SELECT * FROM "giraffes")
name: giraffes
time                 leaves_consumed name   name_1
----                 --------------- ----   ------
2017-01-24T22:54:00Z 12              chrono chrono
2017-01-24T22:54:05Z 13              chrono chrono
2017-01-24T22:54:10Z 15              chrono chrono
2017-01-24T22:54:15Z 13              chrono chrono
2017-01-24T22:54:00Z 8               graf   graf
2017-01-24T22:54:05Z 2               graf   graf
2017-01-24T22:54:10Z 4               graf   graf
2017-01-24T22:54:15Z 5               graf   graf
```

#### Expected behavior:
Just: 
```
name: giraffes
time                 leaves_consumed name
----                 --------------- ----
2017-01-24T22:54:00Z 12              chrono
2017-01-24T22:54:00Z 8               graf
2017-01-24T22:54:05Z 13              chrono
2017-01-24T22:54:05Z 2               graf
2017-01-24T22:54:10Z 15              chrono
2017-01-24T22:54:10Z 4               graf
2017-01-24T22:54:15Z 13              chrono
2017-01-24T22:54:15Z 5               graf
```

#### Actual behavior:

I get a duplicate column of the `name` tag (`name_1`) and it seems like the points are ordered first by series and then by timestamp.
I can get that order issue by selecting specific field keys as well:
```
> SELECT "leaves_consumed" FROM (SELECT "leaves_consumed" FROM "giraffes")
name: giraffes
time                 leaves_consumed
----                 ---------------
2017-01-24T22:54:00Z 12
2017-01-24T22:54:05Z 13
2017-01-24T22:54:10Z 15
2017-01-24T22:54:15Z 13
2017-01-24T22:54:00Z 8
2017-01-24T22:54:05Z 2
2017-01-24T22:54:10Z 4
2017-01-24T22:54:15Z 5
```

### 4. How `GROUP BY time()` vs `GROUP BY tag` works in the main query
(This is more of a comment on subquery behavior that, if anyone's interested, I'd like to discuss/learn more about)

#### Steps to reproduce:

1. Write some data
```
> insert giraffes,name=chrono leaves_consumed=12 1485298440000000000
> insert giraffes,name=graf leaves_consumed=8 1485298440000000000
> insert giraffes,name=chrono leaves_consumed=13 1485298445000000000
> insert giraffes,name=graf leaves_consumed=2 1485298445000000000
> insert giraffes,name=chrono leaves_consumed=15 1485298450000000000
> insert giraffes,name=graf leaves_consumed=4 1485298450000000000
> insert giraffes,name=chrono leaves_consumed=13 1485298455000000000
> insert giraffes,name=graf leaves_consumed=5 1485298455000000000

> select * from giraffes
name: giraffes
time                 leaves_consumed name
----                 --------------- ----
2017-01-24T22:54:00Z 12              chrono
2017-01-24T22:54:00Z 8               graf
2017-01-24T22:54:05Z 13              chrono
2017-01-24T22:54:05Z 2               graf
2017-01-24T22:54:10Z 15              chrono
2017-01-24T22:54:10Z 4               graf
2017-01-24T22:54:15Z 13              chrono
2017-01-24T22:54:15Z 5               graf
```

2. Run a query with a `GROUP BY tag` clause in the main query
```
> SELECT * FROM (SELECT MAX("leaves_consumed") FROM "giraffes" WHERE time >= '2017-01-24T22:54:00Z' AND time <= '2017-01-24T22:54:15Z') GROUP BY "name"
name: giraffes
tags: name=graf
time                 max
----                 ---
2017-01-24T22:54:00Z 8

name: giraffes
tags: name=chrono
time                 max
----                 ---
2017-01-24T22:54:10Z 15
```
3. Run the same query but replace the `GROUP BY tag` clause with a `GROUP BY time()` clause
```
> SELECT * FROM (SELECT MAX("leaves_consumed") FROM "giraffes" WHERE time >= '2017-01-24T22:54:00Z' AND time <= '2017-01-24T22:54:15Z') GROUP BY time(10s)
ERR: error parsing query: GROUP BY requires at least one aggregate function
```

#### Actual behavior and Expected behavior

I know this is how subqueries work now. So a `GROUP BY tag` clause in the main query is automatically applied to the subquery, and a `GROUP BY time()` clause in the main query is not automatically applied to the subquery - but, that difference in behavior for the same InfluxQL clause seems strange to me.

I'd expect that the subquery is basically independent from the main query - so the system runs the subquery and then runs the main query on the data that are returned from the subquery - and no clauses are transported between the two queries. That might be naiive but that's how I thought they'd work.

**Comment by rkuchan** [comment]
@desa - here's the output from our talk today!

**Comment by jsternberg** [comment]
@rkuchan can you split this into three separate tickets and assign them all to me? I think I can respond to them, but it's a bit of an information overload with them all in the same thread. Thanks.

**Comment by jsternberg** [comment]
I figured out the third one. I'll have a PR up for it soon.

**Comment by jsternberg** [comment]
For the fourth one, the reasoning there does make sense. When attempting to validate a query, it validates whether or not that specific query makes sense in its own context. So in this query:

```
SELECT * FROM (SELECT MAX("leaves_consumed") FROM "giraffes" WHERE time >= '2017-01-24T22:54:00Z' AND time <= '2017-01-24T22:54:15Z') GROUP BY time(10s)
```

The `GROUP BY time(10s)` doesn't make sense for a raw query. While it makes sense for the inner query and would be propagated inwards, it doesn't make sense for the query it is attached to and is rejected as a malformed query. Tags don't have that restriction though since grouping by a tag makes sense in both contexts.

But, to make things easier, certain queries will ignore the propagated interval to avoid validation failing in ways that doesn't make sense. So if you have this query:

```
SELECT mean(value) FROM (SELECT 1 / value FROM cpu) WHERE time > now() - 10m GROUP BY host, time(10s)
```

The inner query is a raw query so it doesn't accept grouping by time, but technically the time gets propagated internal within the query engine (and is ignored). So if you have another nested query inside of that with an aggregate, the interval will be propagated through the middle query. But, if you try this:

```
SELECT mean(value) FROM (SELECT 1 / value FROM cpu GROUP BY time(10s)) WHERE time > now() - 10m GROUP BY host, time(10s)
```

This will fail because the query doesn't make sense.

Basically, it fails when the grouping is declared explicitly and it's not compatible with the current query level. It ignores that failure if the grouping is implicit from a level above it.

**Comment by rossmcdonald** [comment]
@rkuchan I've created separate issues for each of your points, so I'm going to go ahead and close out this one.

@jsternberg I've assigned you to all four issues. Please feel free to close any that have already been resolved. Please also open issues for any outstanding subquery issues that do not have GH issues created yet and CC me on them.

**Comment by rkuchan** [comment]
Thanks, @rossmcdonald!
